src/udev/hidraw_id/hidraw_id_desc.c

```diff
@@ -0,0 +1,102 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Modified from "u2f_hidraw_id.c" on 2019-08-20
+ *
+ * Original copyright:
+ *
+ * Copyright (c) 2014-2015 Andrew Lutomirski
+ */
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "hidraw_id_desc.h"
+
+#define HID_RPTDESC_FIRST_BYTE_LONG_ITEM 0xfeu
+#define HID_RPTDESC_TYPE_GLOBAL 0x1u
+#define HID_RPTDESC_TYPE_LOCAL 0x2u
+#define HID_RPTDESC_TAG_USAGE_PAGE 0x0u
+#define HID_RPTDESC_TAG_USAGE 0x0u
+
+/*
+ * HID usage for FIDO U2F/CTAP security tokens.
+ * https://fidoalliance.org/specs/fido-u2f-v1.0-ps-20141009/fido-u2f-u2f_hid.h-v1.0-ps-20141009.txt
+ * https://fidoalliance.org/specs/fido-v2.0-ps-20190130/fido-client-to-authenticator-protocol-v2.0-ps-20190130.html#usb-discovery
+ * https://www.usb.org/sites/default/files/hutrr48.pdf
+ */
+#define FIDO_FULL_USAGE_CTAPHID 0xf1d00001u
+
+/*
+ * Parses a HID report descriptor and identifies HID devices based on their declared usages.
+ * https://www.usb.org/sites/default/files/documents/hid1_11.pdf (Section 6.2.2)
+ */
+int parse_hid_report_desc(const uint8_t *desc, size_t desc_len, struct HidrawIdDescResult *ret) {
+        uint32_t usage;
+        struct HidrawIdDescResult local_ret = {};
+
+        for (size_t pos = 0; pos < desc_len; ) {
+                uint8_t tag, type, size_code;
+                size_t size;
+                uint32_t value;
+
+                /* Report descriptors consists of short items (1-5 bytes) and long items (3-258 bytes). */
+                if (desc[pos] == HID_RPTDESC_FIRST_BYTE_LONG_ITEM) {
+                        /* No long items are defined in the spec; skip them.
+                         * The length of the data in a long item is contained in the byte after the long
+                         * item tag. The header consists of three bytes: special long item tag, length,
+                         * actual tag. */
+                        if (pos + 1 >= desc_len)
+                                return -EINVAL;
+                        pos += desc[pos + 1] + 3;
+                        continue;
+                }
+
+                /* The first byte of a short item encodes tag, type and size. */
+                tag = (uint8_t) desc[pos] >> 4; /* Bits 7 to 4 */
+                type = (uint8_t) ((uint8_t) desc[pos] >> 2) & 0x3; /* Bits 3 and 2 */
+                size_code = (uint8_t) desc[pos] & 0x3; /* Bits 1 and 0 */
```

Comment:
I agree that these ones are, but the ones further down are not, even though they are superficially very similar. My general policy now that I touch the original parser code would be to add casts after any bitwise operation on unsigned types, in order to prevent the hard-to-spot bugs that come with them (like the one that was found by UBsan in an earlier version of this PR). But if you prefer I could think more about which of these casts are essential and which ones aren't, and remove some of them.